src/ucs/stats/stats.c

```diff
@@ -460,6 +477,184 @@ void ucs_stats_node_free(ucs_stats_node_t *node)
+    uint32_t cnt_agrgt_sum_index;
+
+    pthread_mutex_lock(&ucs_stats_context.lock);
+
+    if (ucs_likely(clsid != 0) && (ucs_stats_context.aggrgt_sum_db)) {
+        offset = ucs_stats_context.aggrgt_sum_db[clsid]
+                         .aggregate_sum_class_offset;
+    } else {
+        size_t total_alloc_size_bytes;
+
+        if (ucs_stats_context.aggrgt_sum_db_size >=
+            ucs_stats_context.aggrgt_sum_db_alloc_size) {
+            ucs_stats_context.aggrgt_sum_db_alloc_size +=
+                    UCS_STATS_AGG_SUM_COUNTERS_REALLOC_SZ;
+
+            /* Reallocate clsid->name db */
+            total_alloc_size_bytes =
+                    (ucs_stats_context.aggrgt_sum_db_alloc_size *
+                     sizeof(*ucs_stats_context.aggrgt_sum_db));
+            ucs_stats_context.aggrgt_sum_db = ucs_realloc(
+                    ucs_stats_context.aggrgt_sum_db, total_alloc_size_bytes,
+                    (const char*)UCS_STATS_CLSID_TO_NAME_DB_NAME);
+        }
+
+        /* Initialize entry */
+        ucs_snprintf_safe(
+                ucs_stats_context
+                        .aggrgt_sum_db[ucs_stats_context.aggrgt_sum_db_size]
+                        .class_name,
+                sizeof(ucs_stats_context
+                               .aggrgt_sum_db[ucs_stats_context.aggrgt_sum_db_size]
+                               .class_name),
+                "%s", class_name);
+        ucs_stats_context.aggrgt_sum_db[ucs_stats_context.aggrgt_sum_db_size]
+                .aggregate_sum_class_offset =
+                ucs_stats_context.aggrgt_sum_db_last_offset;
+
+        /* Initialize return value */
+        offset = ucs_stats_context.aggrgt_sum_db_last_offset;
+
+        /* Return class_id */
+        *class_id = ucs_stats_context.aggrgt_sum_db_size;
+
+        /* Update global offset */
+        ucs_stats_context.aggrgt_sum_db_last_offset += num_counters_in_class;
+
+        /* Reallocate counters names db */
+        ucs_stats_context.aggrgt_cnt_names_db_size += num_counters_in_class;
+        total_alloc_size_bytes =
+                (ucs_stats_context.aggrgt_cnt_names_db_size *
+                 sizeof(*ucs_stats_context.aggrgt_cnt_names_db));
+        ucs_stats_context.aggrgt_cnt_names_db = ucs_realloc(
+                ucs_stats_context.aggrgt_cnt_names_db, total_alloc_size_bytes,
+                (const char*)UCS_STATS_AGGRGT_SUM_CNT_NAMES_DB);
+
+        cnt_agrgt_sum_index = offset;
+        ucs_for_each_bit(counter_index, filter_node->counters_bitmask) {
+            ucs_list_for_each(temp_node, &filter_node->type_list_head,
+                              type_list) {
+                ucs_stats_context.aggrgt_cnt_names_db[cnt_agrgt_sum_index].name =
+                        node->cls->counter_names[counter_index];
+                ucs_stats_context.aggrgt_cnt_names_db[cnt_agrgt_sum_index]
+                        .class_name = node->cls->name;
+            }
+            cnt_agrgt_sum_index++;
+        }
```

Comment:
maybe instead of keeping a global list of names we could generate it during the call to ucs_stats_aggregate_sum_names_db_get ?

Reply:
does Score-P actually read the counter names every time, or just their values? is it possible to read names during initialization?